Add HTTP API functionality to use TaskManager as web application

codeofcthulhu opened the pull request:
Create HttpTaskServer to manage http connection and requests/responses. Create Http*Handler classes to handle the majority of task manager methods. Add Unit test for all new http server functionality. Create new Test classes. Five for all Http*Handler classes and one abstract with common to all https://github.com/beforeeach and @AfterEach methods. Also create TypeToken classes for Task, SubTask and Epic Lists for deserialization.

codeofcthulhu committed:
feat: add HTTP API to use TaskManager as web application

Create HttpTaskServer to manage http connection and requests/responses. Create Http*Handler classes to handle the majority of task manager methods.

codeofcthulhu committed:
test: Add Unit test for all new http server functionality

Create new Test classes. Five for all Http*Handler classes and one abstract with common to all @BeforeEach and @AfterEach methods. Also create TypeToken classes for Task, SubTask and Epic Lists for deserialization.

codeofcthulhu committed:
style: code must match code style

Remove unused imports and rename "EpicById" variable

codeofcthulhu committed:
style: code must match code style

Remove unused imports

codeofcthulhu committed:
fix: relocate assertEquals from shouldReturnEmptyPrioritizedList to shouldReturnPrioritizedListWithOneElement

Checking for only one task in the prioritized list accidentally got into the wrong test method.

codeofcthulhu committed:
fix: change "== null" to Objects.isNull()

codeofcthulhu committed:
style: code must match code style

rename "EpicById" variable

AlexeyCheuzov commented on src/http/handlers/BaseHttpHandler.java:
```diff
@@ -0,0 +1,38 @@
+public abstract class BaseHttpHandler implements HttpHandler {
+
+    protected TaskManager manager;
+    protected Gson jsonMapper;
```
Gson можно сделать статическим и использовать во всех потомках

AlexeyCheuzov commented on src/http/handlers/BaseHttpHandler.java:
```diff
@@ -0,0 +1,38 @@
+
+    protected void sendError(HttpExchange exchange, String message, int code) throws IOException {
+        ErrorResponse errorResponse = new ErrorResponse(exchange.getRequestURI().getPath(), message, code
+        );
```
Не совсем корректно переносить одну строку

AlexeyCheuzov commented on src/http/handlers/BaseHttpHandler.java:
```diff
@@ -0,0 +1,38 @@
+        exchange.getResponseHeaders().add("Content-Type", "application/json;charset=utf-8");
+        exchange.sendResponseHeaders(code, response.length);
+
+        try (OutputStream os = exchange.getResponseBody()) {
```
Хорошая практика

AlexeyCheuzov commented on src/http/handlers/HttpEpicHandler.java:
```diff
@@ -0,0 +1,141 @@
+        String method = exchange.getRequestMethod();
+        try {
+            switch (method) {
+                case "GET":
```
Названия методов лучше вынести в константу, как и коды ответов

AlexeyCheuzov commented on src/http/server/HttpTaskServer.java:
```diff
@@ -0,0 +1,90 @@
+    }
+
+    public void start() throws IOException {
+        InetSocketAddress address = new InetSocketAddress("localhost", 8080);
```
Конфигурационные данные, такие как номер порта или хост, также пока лучше вынести в константу.

codeofcthulhu committed:
refactor: put some fields in constants or static fields

JsonMapper and manager should be static fileds, because its are common for all BaseHttpHandler subclasses. To set this fields you should call setUpHandlers(Gson j, Task manager m) method in start() method of HttpTaskServer class. Http request methods should stored as enum, as well as Http Server response codes.
